packages/plugins/typescript/typescript/tests/typescript.spec.ts

```diff
@@ -2536,6 +2536,79 @@ describe('TypeScript', () => {
       `);
       validateTs(result);
     });
+
+    describe('@oneOf on input types', () => {
+      const oneOfDirectiveDefinition = /* GraphQL */ `
+        directive @oneOf on INPUT_OBJECT
+      `;
+
+      it('correct output for type with single field', async () => {
+        const schema = buildSchema(
+          /* GraphQL */ `
+          input Input @oneOf {
+            int: Int
+          }
+
+          type Query {
+            foo(input: Input!): Boolean!
+          }
+        `.concat(oneOfDirectiveDefinition)
+        );
+
+        const result = await plugin(schema, [], {}, { outputFile: '' });
+
+        expect(result.content).toBeSimilarStringTo(`
+          export type Input =
+            { int?: InputMaybe<Scalars['Int']>; };
+        `);
+      });
+
+      it('correct output for type with multiple fields', async () => {
+        const schema = buildSchema(
+          /* GraphQL */ `
+          input Input @oneOf {
+            int: Int
+            boolean: Boolean
+          }
+
+          type Query {
+            foo(input: Input!): Boolean!
+          }
+        `.concat(oneOfDirectiveDefinition)
+        );
+
+        const result = await plugin(schema, [], {}, { outputFile: '' });
+
+        expect(result.content).toBeSimilarStringTo(`
+          export type Input =
+            { int?: InputMaybe<Scalars['Int']>; }
+            | { boolean?: InputMaybe<Scalars['Boolean']>; };
+        `);
+      });
+
+      it('correct output for type with non-optional fields', async () => {
+        const schema = buildSchema(
+          /* GraphQL */ `
+          input Input @oneOf {
+            int: Int!
+            boolean: Boolean!
+          }
+
+          type Query {
+            foo(input: Input!): Boolean!
+          }
+        `.concat(oneOfDirectiveDefinition)
+        );
+
+        const result = await plugin(schema, [], {}, { outputFile: '' });
+
+        expect(result.content).toBeSimilarStringTo(`
+          export type Input =
+            { int: Scalars['Int']; }
+            | { boolean: Scalars['Boolean']; };
```

Comment:
This is accurate, but I think it'll be uncomfortable to work with, because you'll have to do something like `if ('int' in foo)` rather than `if (foo.int)`.

Maybe something like:

```
          export type Input =
            | { int: Scalars['Int']; boolean?: never }
            | { boolean: Scalars['Boolean']; int?: never };
```

https://www.typescriptlang.org/play?#code/C4TwDgpgBAygxgQwDYIE4GcoF4oG8BQURUAkgHbABcUZArgLYBGEqA3IcQEID23SECMtUa9+g9gF98+CAA8w3VMCihIpMmFrKsHAD54oASwrV4yNOgDaAcnLBrAXVZQpRfbigi+AobEQoMGx5vQUdnCXYZeUVlVWhyTWAAJmw9A2MqP3NA2wowz1EfAH5qMggANxYXNI8vMV8zAKtrYPr8jJKaCqqI6ThuMnRlBGoErWwaWiQkKARMQRB2fsHlRlGNLRScOmnZ+bJF6UMAMygACgQAOgyASjwOImX0b0ukbgBzC+uKG+coAHp-lBOABBAAi+Ck+BO52sGWsRjIszuBGIUCeLzenyut0kR1OZ0Y32AKIe6IGz34rw+hOJv0h+Nh8MRnlJaIxVKxtNxAKBoIhEiAA

Take this suggestion with a pinch of salt; I don't know what the nicest TypeScript would be - it's a tradeoff either way

Reply:
@benjie Yeah I think this is definitely the better solution when dealing with it in your resolver function! So far I only thought of it from a client perspective.

Reply:
@benjie https://github.com/dotansimha/graphql-code-generator/pull/7886/commits/fe6369a6ea945b79c273b2f0fc5c72e15ba2ff8a